### DISCUSSION TITLE
Adventure map, tree of knowledge advances two levels in a row

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#8042] [General]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

I was lucky to have two lvlups from a single tree. :)

https://github.com/ihhub/fheroes2/assets/55348946/6e477db1-7703-4ebf-abf4-a0180760cdeb

### Save file

[Double bonus.zip](https://github.com/ihhub/fheroes2/files/13325437/Double.bonus.zip)

**Comment by Mr-Bajs** [comment]
Also the hero name seems to be ???????, looks like a bug from the video.

"??????? Has gained a level."

**Comment by Branikolog** [comment]
Hi, @Mr-Bajs 

> Also the hero name seems to be ???????, looks like a bug from the video.
> 
> "??????? Has gained a level."

Question marks are shown for some generated map texts when the language of the game is not coincides with the language selected for a particular map. I was just played that map in non-English mode, but switched to English for recording a video. :)

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog this is not a bug, the same happens in the original game in the same circumstances. Your hero has 7324 experience, I set the same amount of experience for the hero on the test map and voila:

https://github.com/ihhub/fheroes2/assets/32623900/5ec354df-8e19-4d32-8bc1-f051cdc5069b

That's because the Tree of Knowledge always awards your hero with the amount of experience that equals to the difference between the amount of experience needed for the current hero level and the amount of experience needed for the next hero level. 7324 exp is level 6, so Tree of Knowledge will give us 7700 (level 7) - 6000 (level 6) = 1700 exp, so we will have 9024 exp, which is level 8 (level 6 + 2). The same happens in the original game.

**Comment by Mr-Bajs** [comment]
Though one could argue that the original also has the bug. It doesnt make sence the way it works. If introducing bugs from the original for the sake of keeping the original game, sort of makes the purpose for fheroes2 invaild. 

Though it depens if one consider this a bug in the og game or design choice. 
I do think in this case is the best way do look at this as a bug or a faulty design from the original game and should be changed and improved in fheroes2.

**Comment by zenseii** [comment]
Hi, @Mr-Bajs.

> [...] faulty design from the original game and should be changed and improved in fheroes2.

I disagree. The Tree of Knowledge doesn't say that it will grant you just 1 level up, it only says that it will grant you some "experience":

![image](https://github.com/ihhub/fheroes2/assets/12501091/68f53294-049d-481a-a2a7-31433408ee39)


The way it works in Heroes 2 is the same as it the trees in later Heroes series games work, unless I'm remembering wrong.

**Comment by Mr-Bajs** [comment]
Okey, if it doesnt clearly stated its one level, then i could agree. I'll guess ill just asumed its +1 level because I always though it was so.

**Comment by oleg-derevenetz** [comment]
In HoMM3, a different [experience scale system](https://heroes.thelazy.net/index.php/Experience) is used, in which more experience is needed to move to next level than was required to move from the previous level to the current one. In HoMM2, this is not always the case. For example, in HoMM2 you need to get 1700 exp to move from level 6 to level 7, but you need just 1300 exp to move from level 7 to level 8. Tree of Knowledge works absolutely the same way in HoMM2 and HoMM3, only the experience scale system differs.

**Comment by Branikolog** [comment]
Hi everyone!

@oleg-derevenetz , thanks for clarification the formula. It was really helpful for me to understand the way it works.

Indeed, in the OG there is no information (I'm not sure about the official guides, though), that this object rises exactly one level.
However in later series there's clear statement, that the tree rises a single level:
![image](https://github.com/ihhub/fheroes2/assets/55348946/b88ebf2e-ec49-4f05-8ad5-f140f514ddaf)
![image](https://github.com/ihhub/fheroes2/assets/55348946/c73fb6be-96e6-4251-b6f9-9d7cdeb9f078)

So I cannot insist on fixing this, since it works according to the original logic and the formula. But I'm quite sure all players (same as me and some of you, guys) are expecting this tree to work the same, as in later series (i.e. adding only 1 level).
According to the experience scale table I found in internet, there seems to be only one case, when a hero can get 2 levels (on a 6th level close to 7th).
So this issue could be considered as unique game feature (which could be abused by players, who know, how this works). But I personally find this behaviour confusing and looking-like-a-bug. 
I can only humbly suggest rounding the amount of experience hero gains in this particular case, so in result he has 8999 points (8lvl - 1 exp point).
But I won't object if some of our team members disagree with such approach.

**Comment by zenseii** [comment]
Hi, @Branikolog! I acknowledge that at first sight gaining 2 levels from the tree seems like a bug. To me it makes perfect sense that it can happen. Personally I would not want to add a specific exception to how much you get only in that case. It would just be even more confusing when in all other times the tree will give you exp according to the rules described by @oleg-derevenetz.

I have a different suggestion, and it would be a chance to do one better than Heroes 3 and later games: How about displaying the exact amount of exp you get in the Tree of Knowledge dialog? Something like +1300 exp.

This would fit very well in with how fheroes2 gives power to the player by exposing how things that aren't explained in the manual work, so that the player can make more informed decisions, like exposing the turn order, numerical movement points, etc.

As has been hinted at in this thread, many players (not saying the people in this thread) falsely believe that the Tree of Knowledge will just give you the exact exp needed for the next level, even if that is just 1 exp. This would help them understand that this is wrong, and indeed wrong for later games too.

**Comment by Branikolog** [comment]
Hi, @zenseii !

>How about displaying the exact amount of exp you get in the Tree of Knowledge dialog? Something like +1300 exp.

Doesn't it work like that currently? :)

>It would just be even more confusing when in all other times the tree will give you exp according to the rules described by @oleg-derevenetz.

The thing is that only 0.01% of players know the exact formula. :)
Even I, having been testing this game for a several years already, don't know, how much exp a hero is going to acquire on the particular level. The majority of players know, that the tree gives experience ~equal to one level. Like a 1-2k of points on the first 10 levels and even more on higher.
Some players believe, if a hero needs 5 exp points to level up, tree gives just 5. But we already have an option to show the amount of exp tree gives:
![image](https://github.com/ihhub/fheroes2/assets/55348946/e75592cd-3f4d-4d23-a02a-55b6192f3185)

**Comment by zenseii** [comment]
@Branikolog,

> Doesn't it work like that currently? :)

Ah yes I completely forgot that! In this case I think it's fine. I expect experienced players to know how the Tree works, but I would only expect very pro players (1 % of players?) to have memorized the experience scale table (I personally only remember the first couple of levels.).

I do, however, believe that the plain displaying of the experience points gained "1700" could be improved. Bigger text? Yellow text? Add a + sign so it says "+1700" ? Add "EXP" after it? After all EXP should be translated differently in different languages anyways so the image above isn't necessarily clear to non-English speakers. "+1700 EXP"

How about adding in parenthesis next to the experience points the number of levels gained: "(+1 level)" or "(+2 levels)" (for that specific case)? It would make it even more clear for players.

"+1700 EXP (+1 Level)"

**Comment by LeHerosInconnu** [comment]
Hello everyone,

The operation of the tree of knowledge is fine, it's the hero's progression that's the problem.
Hey, look at this, I've made a proposal for it: [Progression of the hero's levels](https://github.com/ihhub/fheroes2/discussions/6727). :)

**Comment by Branikolog** [comment]
So, I believe it's much more confusing for players to get 2 lvlups, rather than get less exp points (up to -400. If a hero has 7699 exp and the next level requires 9000, so he gets just 1300 instead of 1700), which they hardly even notice, because noone know the exact values. :)
I think the solution proposed by @LeHerosInconnu is also reasonable: just changing a little exp progression for a couple of levels (6, 7).

It just much easier, than explaining players, why do they get 2 levels instead of 1. :)

**Comment by zenseii** [comment]
Hi, @LeHerosInconnu,

> The operation of the tree of knowledge is fine, it's the hero's progression that's the problem. Hey, look at this, I've made a proposal for it: [Progression of the hero's levels](https://github.com/ihhub/fheroes2/discussions/6727). :)


I didn't want to mention changing the hero level progression, because I do not want to open that can of worms. It touches too much of what is the game's balance that changing it based solely on "I prefer my numbers to be ordered like this.", when there are too many factors concerning the balancing of the game to consider when changing such a core aspect of the game.


EDIT: I also don't think changing the hero level progression, just because "The tree of knowledge suprised me" is enough of an argument. Of course, no one is stopping you from going into the rabbit hole of the balancing of why a Knight needs to get to that level at that point, have x % chance of getting a skill, at what is usually the point of the game when you gain that level, and the creatures you would expect to have at such a level.

Personally, I don't think it's worth the time. Especially considering that the biggest fans of the game will just be negatively surprised when they don't level up when they expected to.

**Comment by Branikolog** [comment]
So I'm converting this issue into discussion, since this is going out of the scope of reproducing the original game.
After all this double lvlup could be a special feature added purposely allowing players to reach the 8th lvl faster. If I'm not mistaken this progression was taken from H1, so there's a high chance, than original developers made it intentionally.

**Reply by zenseii** [reply]
> If I'm not mistaken this progression was taken from H1, so there's a high chance, than original developers made it intentionally.

Indeed the same thing happens in Heroes 1 still from level 7 to 8:
![image](https://github.com/ihhub/fheroes2/assets/12501091/888480cc-1b2e-4ad5-b7f1-524a72ad002e)


So if we think this was a flaw then we have to believe that NWC made this mistake twice. It's just down to subjective opinion because we'll likely never find out the truth about it, and anyways none of it does anything in the form of giving arguments based on actual game design nor provide any statistical data showing that the balance needs to change.

Picture from here: https://mightandmagic.fandom.com/wiki/Experience_(Heroes)#Heroes_of_Might_and_Magic_II
